src/main/java/io/vertx/core/net/KeyStoreOptions.java

```diff
@@ -0,0 +1,150 @@
+/*
+ * Copyright (c) 2011-2019 Contributors to the Eclipse Foundation
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+ * which is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+ */
+
+package io.vertx.core.net;
+
+import io.vertx.codegen.annotations.GenIgnore;
+import io.vertx.core.buffer.Buffer;
+
+/**
+ * Key or trust store options configuring private key and/or certificates based on {@code KeyStore}.
+ *
+ * <ul>
+ *   <li>when used as a key store, it should point to a store containing a private key and its certificate.</li>
+ *   <li>when used as a trust store, it should point to a store containing a list of trusted certificates.</li>
+ * </ul>
+ *
+ * <p> The store can either be loaded by Vert.x from the filesystem:
+ *
+ * <pre>
+ * HttpServerOptions options = HttpServerOptions.httpServerOptions();
+ * options.setKeyStore(new KeyStoreOptions().setType("JKS").setPath("/mykeystore.jks").setPassword("foo"));
+ * </pre>
+ *
+ * Or directly provided as a buffer:
+ *
+ * <pre>
+ * Buffer store = vertx.fileSystem().readFileBlocking("/mykeystore.jks");
+ * options.setKeyStore(new JKSOptions().setType("JKS").setValue(store).setPassword("foo"));
+ * </pre>
+ *
+ * <p> You can also use specific subclasses {@link JksOptions} or {@link PfxOptions} that will set
+ * the {@link #setType} for you:
+ *
+ * <pre>
+ * HttpServerOptions options = HttpServerOptions.httpServerOptions();
+ * options.setKeyStore(new JksOptions().setPath("/mykeystore.jks").setPassword("foo"));
+ * </pre>
+ *
+ * @author <a href="mailto:julien@julienviet.com">Julien Viet</a>
+ * @author <a href="http://tfox.org">Tim Fox</a>
+ */
+public class KeyStoreOptions implements KeyCertOptions, TrustOptions {
+
+  private String type;
```

Comment:
Perhaps we could have a default value for this. For example, on `auth` I was doing:

```java
 // Defaults
  private static final String TYPE = KeyStore.getDefaultType();
```

As this picks the runtime JDK default, for example on Java8 is `JKS` on java9 and up `PKCS12` and maybe on other jdk's or runtimes started with custom security config it will be something else, e.g.: `BKS`

Reply:
that's dangerous I think

Reply:
I mean it is fine to have type null by default and let user set an explicit value, if user does not want to set the value then using JksOptions or PfxOptions is just fine